src/NServiceBus.Core/Pipeline/Incoming/ReceiveFeature.cs

```diff
@@ -20,9 +22,15 @@ protected internal override void Setup(FeatureConfigurationContext context)
             context.Pipeline.Register("TransportReceiveToPhysicalMessageProcessingConnector", b => b.Build<TransportReceiveToPhysicalMessageProcessingConnector>(), "Allows to abort processing the message");
             context.Pipeline.Register("LoadHandlersConnector", b => b.Build<LoadHandlersConnector>(), "Gets all the handlers to invoke from the MessageHandler registry based on the message type.");
 
-            context.Pipeline.Register("ExecuteUnitOfWork", new UnitOfWorkBehavior(), "Executes the UoW");
-            context.Pipeline.Register("MutateIncomingTransportMessage", new MutateIncomingTransportMessageBehavior(), "Executes IMutateIncomingTransportMessages");
-            context.Pipeline.Register("MutateIncomingMessages", new MutateIncomingMessageBehavior(), "Executes IMutateIncomingMessages");
+            var hasUnitsOfWork = context.Container.HasComponent<IManageUnitsOfWork>();
+            context.Pipeline.Register("ExecuteUnitOfWork", new UnitOfWorkBehavior(hasUnitsOfWork), "Executes the UoW");
+
+            var hasIncomingTransportMessageMutators = context.Container.HasComponent<IMutateIncomingTransportMessages>();
```

Comment:
@WilliamBZA TLDR is

currently, the core always registers behaviors and opts out with flags when the behavior effect is not needed
we wanted to stay with that approach instead of dynamically register or not register the behavior
it is also more backward compliant since users could be relying on the behavior being there with the InsertBefore and InsertAfter methods